Objects/stringlib/fastsearch.h

```diff
@@ -324,222 +359,402 @@ typedef struct STRINGLIB(_pre) {
 
 
 static void
-STRINGLIB(_preprocess)(const STRINGLIB_CHAR *needle, Py_ssize_t len_needle,
-                       STRINGLIB(prework) *p)
+STRINGLIB(_init_bc_table_gs_gap)(STRINGLIB(prework) *pw, int dir)
 {
-    p->needle = needle;
-    p->len_needle = len_needle;
-    p->cut = STRINGLIB(_factorize)(needle, len_needle, &(p->period));
-    assert(p->period + p->cut <= len_needle);
-    p->is_periodic = (0 == memcmp(needle,
-                                  needle + p->period,
-                                  p->cut * STRINGLIB_SIZEOF_CHAR));
-    if (p->is_periodic) {
-        assert(p->cut <= len_needle/2);
-        assert(p->cut < p->period);
+    // 1. Fill up a compressed Boyer-Moore "Bad Character" table
+    const STRINGLIB_CHAR *p = pw->p;
+    Py_ssize_t m = pw->m;
+    Py_ssize_t stt = dir == 1 ? 0 : m - 1;
+    Py_ssize_t end = dir == 1 ? m - 1 : 0;
+    Py_ssize_t not_found_shift = Py_MIN(m, MAX_SHIFT);
+    for (Py_ssize_t i = 0; i < (Py_ssize_t)TABLE_SIZE; i++) {
+        pw->table[i] = Py_SAFE_DOWNCAST(not_found_shift,
+                                        Py_ssize_t, SHIFT_TYPE);
     }
-    else {
-        // A lower bound on the period
-        p->period = Py_MAX(p->cut, len_needle - p->cut) + 1;
+    for (Py_ssize_t i = m - not_found_shift; i < m; i++) {
+        SHIFT_TYPE shift = Py_SAFE_DOWNCAST(m - 1 - i,
+                                            Py_ssize_t, SHIFT_TYPE);
+        pw->table[p[stt + dir*i] & TABLE_MASK] = shift;
     }
-    // The gap between the last character and the previous
-    // occurrence of an equivalent character (modulo TABLE_SIZE)
-    p->gap = len_needle;
-    STRINGLIB_CHAR last = needle[len_needle - 1] & TABLE_MASK;
-    for (Py_ssize_t i = len_needle - 2; i >= 0; i--) {
-        STRINGLIB_CHAR x = needle[i] & TABLE_MASK;
+    // 2. Initialize "Good Suffix" Last Character Gap
+    // Note: gap("___aa") = 1
+    pw->gap = m;
+    STRINGLIB_CHAR last = p[end] & TABLE_MASK;
+    for (Py_ssize_t i = 1; i < m; i++) {
+        STRINGLIB_CHAR x = p[end - dir*i] & TABLE_MASK;
         if (x == last) {
-            p->gap = len_needle - 1 - i;
+            pw->gap = i;
             break;
         }
     }
-    // Fill up a compressed Boyer-Moore "Bad Character" table
-    Py_ssize_t not_found_shift = Py_MIN(len_needle, MAX_SHIFT);
-    for (Py_ssize_t i = 0; i < (Py_ssize_t)TABLE_SIZE; i++) {
-        p->table[i] = Py_SAFE_DOWNCAST(not_found_shift,
-                                       Py_ssize_t, SHIFT_TYPE);
+    LOG("Good Suffix Gap: %ld\n", pw->gap);
+}
+
+
+static void
+STRINGLIB(_init_critical_fac)(STRINGLIB(prework) *pw, int dir)
+{
+    // Calculate Critical Factorization
+    const STRINGLIB_CHAR *p = pw->p;
+    Py_ssize_t m = pw->m;
+    Py_ssize_t cut, period;
+    int is_periodic;
+    cut = STRINGLIB(_factorize)(p, m, &period, dir);
+    assert(cut + period <= m);
+    if (dir == 1) {
+        is_periodic = 0 == memcmp(p, p + period, cut * STRINGLIB_SIZEOF_CHAR);
```

Comment:
```suggestion
        is_periodic = memcmp(p, p + period, cut * STRINGLIB_SIZEOF_CHAR) == 0;
```

Just for clarity.

Reply:
If necessary you can also add parentheses:

```suggestion
        is_periodic = (memcmp(p, p + period, cut * STRINGLIB_SIZEOF_CHAR) == 0);
```